Dev Loop Snake

JNotelddim opened the pull request:
Just adding the ability for the snake to follow its own tail and stuff. 
Am now realizing I didn't write tests to go along with it :v: woops :P

JNotelddim committed:
Follow tail

JNotelddim committed:
Circle perimeter

go around perimeter of board while not getting to food.
I've found this leads to snake dying much faster multiple snakes on the board
compared to alternatives with just following tail.

mattcasiro commented on src/brain.py:
```diff
@@ -86,3 +93,144 @@ def get_moves_for_nearest_food(self) -> Optional[List[str]]:
+
+    def follow_tail(self) -> List[str]:
+        """Get moves for getting to tail."""
+        tail = self.me.coordinates[len(self.me.coordinates) - 1]
```
- [x] `self.me.coordinates[-1]`

mattcasiro commented on src/brain.py:
```diff
@@ -20,10 +22,15 @@ def get_decision(self) -> str:
         valid_moves = self.get_valid_moves()
 
         if not valid_moves:
-            return "left"
```
Can we comment the logic here of why we're handling 'there's no valid move' in this way?

- [x] Add comment

JNotelddim committed:
Add comment explaining removal of default left move

JNotelddim committed:
Update brain test

JNotelddim committed:
Accept risk of headons when there are no other move options

Still wanna avoid headon collisions in general, but there's a case which
I've seen several times now where the snake considered the
headon-collisions-risk-square an invalid move, so it goes towards its
tail as default because there are not valid_moves. but if it went
towards the headon in those particular cases, it could've survived.

from running a few games it seems rare enough that it only occurs like 4-5
 times in a 1500-turn game with 6 snakes. so not a huge issue, but also
worth avoiding in my opinion.

mattcasiro commented on app/api.py:
```diff
@@ -21,8 +21,8 @@ def start_response(color):
 def move_response(move):
-    assert move in ['up', 'down', 'left', 'right'], \
-        "Move must be one of [up, down, left, right]"
+#    assert move in ['up', 'down', 'left', 'right'], \
```
Are we sure about getting rid of this? Is this for better passthrough of errors to the unit test output? Maybe we should comment the purpose of removing it so we remember the reasoning.

JNotelddim replied:
err idk that was you on my computer /shrug

mattcasiro commented on src/brain.py:
```diff
@@ -47,7 +54,7 @@ def get_valid_moves(self) -> List[str]:
     def get_nearest_food(self) -> Optional[Coordinate]:
         """Get the food item which has coordinates closest to this snake's head."""
-        closest_food = (Coordinate((0,0)), 9999.0)
+        closest_food = (Coordinate((0,0), None), 9999.0)
```
I think this will need to be reverted based on the changes to `Coordinate` in the other PR

- [x] Remove `None`

mattcasiro reviewed (changes requested):
Love the way you set up the movement algorithms!

JNotelddim committed:
Update syntax in follow_tail

JNotelddim committed:
Undo mistakenly commited commented code

mattcasiro reviewed (approved):
Looks good!
